src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs

```diff
@@ -410,267 +584,458 @@ private string GenerateForObject(TypeMetadata typeMetadata)
+                Debug.Assert(!(
+                    _currentContext.TypesWithMetadataGenerated.Count == 0
+                    && _currentContext.RuntimePropertyNames.Count > 0));
 
-            private static string GetUsingStatementsString(HashSet<string> usingStatements)
-            {
-                string[] usingsArr = usingStatements.ToArray();
-                Array.Sort(usingsArr);
-                return string.Join("\n", usingsArr);
-            }
-
-            private static HashSet<string> GetUsingStatements(TypeMetadata typeMetadata)
-            {
-                HashSet<string> usingStatements = new();
-
-                // Add library usings.
-                usingStatements.Add(FormatAsUsingStatement("System.Runtime.CompilerServices"));
-                usingStatements.Add(FormatAsUsingStatement("System.Text.Json"));
-                usingStatements.Add(FormatAsUsingStatement("System.Text.Json.Serialization"));
-                usingStatements.Add(FormatAsUsingStatement("System.Text.Json.Serialization.Metadata"));
+                StringBuilder sb = new();
 
-                // Add imports to root type.
-                usingStatements.Add(FormatAsUsingStatement(typeMetadata.Type.Namespace));
+                sb.Append(@$"{GetJsonContextDeclarationSource()}
+{{");
 
-                switch (typeMetadata.ClassType)
+                foreach (string propName in _currentContext.RuntimePropertyNames)
                 {
-                    case ClassType.Nullable:
-                        {
-                            AddUsingStatementsForType(typeMetadata.NullableUnderlyingTypeMetadata!);
-                        }
-                        break;
-                    case ClassType.Enumerable:
-                        {
-                            AddUsingStatementsForType(typeMetadata.CollectionValueTypeMetadata);
-                        }
-                        break;
-                    case ClassType.Dictionary:
-                        {
-                            AddUsingStatementsForType(typeMetadata.CollectionKeyTypeMetadata);
-                            AddUsingStatementsForType(typeMetadata.CollectionValueTypeMetadata);
-                        }
-                        break;
-                    case ClassType.Object:
-                        {
-                            if (typeMetadata.PropertiesMetadata != null)
-                            {
-                                foreach (PropertyMetadata metadata in typeMetadata.PropertiesMetadata)
-                                {
-                                    AddUsingStatementsForType(metadata.TypeMetadata);
-                                }
-                            }
-                        }
-                        break;
-                    default:
-                        break;
+                    sb.Append($@"
+    private static {JsonEncodedTextTypeRef} {propName}PropName = {JsonEncodedTextTypeRef}.Encode(""{propName}"");");
```

Comment:
Nice optimization of private bytes, startup and throughput by usage of single list of shared property names + JsonEncodedText.